### PR TITLE
Openshift template support

### PR DESCRIPTION
**Pull request by phracek**
This Pull Request brings a new functionality, how to testing containers via OpenShift templates.

- [x] MTF is able to testing containers via OpenShift templates
- [x] documentation is updated.

### CODE REVIEW

**Comment by phracek** [comment]
Old PR based on devel branch is https://github.com/fedora-modularity/meta-test-family/pull/203

File: .gitignore [hunk 1]
```diff
@@ -97,4 +97,6 @@ ENV/
 # Generated files by framework
 examples/baseruntime/generated.py
 examples/memcached/generated.py
+downloaded_git
```
Comment (TomasTomecek) [inline .gitignore]: you can use `.git/info/exclude` to ignore your local files, no need to store those in repo-wide .gitignore

File: examples/openshift/template/memcached.py [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+from moduleframework import common
+import time
+
+
+class SanityCheck1(module_framework.AvocadoTest):
+    """
+    :avocado: enable
+    """
+
+    def test_sanity(self):
```
Comment (TomasTomecek) [inline examples/openshift/template/memcached.py]: would be nice to use a more complex test where you actually connect to memcached

File: moduleframework/common.py [hunk 3]
```diff
@@ -93,6 +93,9 @@
 OPENSHIFT_INIT_WAIT = 50
 STATIC_LINTERS = 'static'
 GENERIC_TEST = 'generic'
+OPENSHIFT_DOCKER_REGISTER = "docker-registry"
```
Comment (TomasTomecek) [inline moduleframework/common.py]: `s/REGISTER/REGISTRY/`

File: moduleframework/common.py [hunk 4]
```diff
@@ -461,6 +464,13 @@ def get_url(self):
         """
         return self.info.get("url")
 
+    def get_template(self):
+        """
+        get location of template
+        :return:
```
Comment (TomasTomecek) [inline moduleframework/common.py]: so, what it should return?

File: moduleframework/helpers/openshift_helper.py [hunk 5]
```diff
@@ -54,8 +58,52 @@ def __init__(self):
         # application name is taken from docker.io/modularitycontainer/memcached
         self.app_name = self.container_name.split('/')[-1]
         self.app_ip = None
+        self.project_name = "%s-%s" % (''.join(random.choice(string.lowercase) for _ in range(3)), self.app_name)
```
Comment (TomasTomecek) [inline moduleframework/helpers/openshift_helper.py]: I would actually do it the other way around: `{project_name}-{random_str}` -- you'll have nice grouping once sorting alphabetically

File: moduleframework/helpers/openshift_helper.py [hunk 6]
```diff
@@ -54,8 +58,52 @@ def __init__(self):
...
         common.print_debug(self.icontainer, self.app_name)
 
+    def _get_openshift_ip_registry(self):
+        openshift_ip_register = None
+        docker_registry = self.runHost('oc get svc -n default -o json', ignore_status=True).stdout
+        docker_registry = self._convert_string_to_json(docker_registry)
+        for reg in docker_registry:
```
Comment (TomasTomecek) [inline moduleframework/helpers/openshift_helper.py]: you don't need this loop, you can just do `oc get svc docker-registry`

File: moduleframework/helpers/openshift_helper.py [hunk 7]
```diff
@@ -54,8 +58,52 @@ def __init__(self):
...
+        if password is None:
+            s = self.runHost("oc login -u %s" % account, verbose=common.is_not_silent())
+        else:
+            s = self.runHost("oc login -u %s -p %s" % (account, password), verbose=common.is_not_silent())
+
+
+    def _register_docker_to_openshift_register(self):
+        whoami = self.runHost("oc whoami -t", ignore_status=True, verbose=common.is_not_silent()).stdout
+        self._change_openshift_account()
+        self.runHost('docker pull %s' % self.container_name)
```
Comment (TomasTomecek) [inline moduleframework/helpers/openshift_helper.py]: why pulling it? what if I built the image locally

File: moduleframework/helpers/openshift_helper.py [hunk 8]
```diff
@@ -54,8 +58,52 @@ def __init__(self):
...
+        openshift_ip_register = self._get_openshift_ip_registry()
+        self._change_openshift_account(account=common.get_openshift_user(),
+                                password=common.get_openshift_passwd())
+
+        for i in range(0,5):
+            docker_login = self.runHost('docker login -u developer -p %s %s:5000' % (whoami, openshift_ip_register))
```
Comment (TomasTomecek) [inline moduleframework/helpers/openshift_helper.py]: "developer" is hardcoded - what if someone deploys openshift without that account?

File: moduleframework/helpers/openshift_helper.py [hunk 9]
```diff
@@ -66,8 +114,7 @@ def _app_exists(self):
         if int(oc_status.exit_status) == 0:
             common.print_info("Application already exists.")
             return True
-        oc_services = self.runHost("oc get pods -o json", ignore_status=True).stdout
-        oc_services = self._convert_string_to_json(oc_services)
+        oc_services = self._oc_get_output('pods')
```
Comment (TomasTomecek) [inline moduleframework/helpers/openshift_helper.py]: shouldn't this be named `oc_pods`?

File: moduleframework/helpers/openshift_helper.py [hunk 10]
```diff
@@ -97,6 +144,27 @@ def _check_app_in_json(self, item):
         except KeyError:
             return False
 
+    def _check_template_in_json(self, item):
+        """
+        Function checks if json_output contains container with specified name
+
+
+        :param json_output: json output from an OpenShift command
```
Comment (TomasTomecek) [inline moduleframework/helpers/openshift_helper.py]: there is no such parameter

File: moduleframework/helpers/openshift_helper.py [hunk 11]
```diff
@@ -97,6 +144,27 @@ def _check_app_in_json(self, item):
...
+                 False if the application does not exist
+        """
+        try:
+            try:
+                if item.get('kind') == "Template":
+                    if item.get('metadata').get('name') == self.app_name:
+                        return True
+                else:
+                    return False
+            except AttributeError:
```
Comment (TomasTomecek) [inline moduleframework/helpers/openshift_helper.py]: you can actually do
```python
except (Exception, OtherException) as ex:
```

File: moduleframework/helpers/openshift_helper.py [hunk 12]
```diff
@@ -109,43 +177,109 @@ def _convert_string_to_json(self, inp_string):
...
+    def _get_openshift_template(self):
+        template_name = self._oc_get_output('template')
+        common.print_debug(template_name)
+
+    def _oc_get_output(self, namespace):
         """
-        It removes an application from specific "namespace" like svc, dc, is.
-        :param oc_service: Service from which we would like to remove application
+        Function returns json output for specific namespace
+        :param namespace:
```
Comment (TomasTomecek) [inline moduleframework/helpers/openshift_helper.py]: yes?

File: moduleframework/helpers/openshift_helper.py [hunk 13]
```diff
@@ -109,43 +177,109 @@ def _convert_string_to_json(self, inp_string):
...
+        :return:
```
Comment (TomasTomecek) [inline moduleframework/helpers/openshift_helper.py]: same here

File: moduleframework/helpers/openshift_helper.py [hunk 14]
```diff
@@ -109,43 +177,109 @@ def _convert_string_to_json(self, inp_string):
...
         """
         # Check status of svc/dc/is
-        oc_get = self.runHost("oc get %s -o json" % oc_service, ignore_status=True).stdout
+        oc_get = self.runHost("oc get %s -o json" % namespace, ignore_status=True).stdout
```
Comment (TomasTomecek) [inline moduleframework/helpers/openshift_helper.py]: I think you mean `s/namespace/resource/`.

File: moduleframework/helpers/openshift_helper.py [hunk 15]
```diff
@@ -109,43 +177,109 @@ def _convert_string_to_json(self, inp_string):
...
         oc_get = self._convert_string_to_json(oc_get)
-        # The output is like
-        # dovecot     172.30.1.1:5000/myproject/dovecot     latest    15 minutes ago
-        # memcached   172.30.1.1:5000/myproject/memcached   latest    13 minutes ago
+        return oc_get
 
+    def _remove_apps_from_openshift_namespaces(self, oc_service="svc"):
+        """
+        It removes an application from specific "namespace" like svc, dc, is.
```
Comment (TomasTomecek) [inline moduleframework/helpers/openshift_helper.py]: see above, those are not namespaces, but resources:

```
$ oc get
You must specify the type of resource to get. Valid resource types include:

    * all
    * buildconfigs (aka 'bc')
    * builds
    * certificatesigningrequests (aka 'csr')
    * clusters (valid only for federation apiservers)
    * clusterrolebindings
```

File: moduleframework/helpers/openshift_helper.py (hunk 7 above)
Comment (phracek) [reply moduleframework/helpers/openshift_helper.py]: Good point. It can be specified in config.yaml file like `docker_pull: False` which will suppress it.

File: docs/user_guide/how_to_write_conf_file.rst [hunk 16]
```diff
@@ -62,16 +66,18 @@ An example of module types specification:
...
-* **container** contains a link to a container (docker.io or local tar.gz file) (obsolote)
-* **repo** is used when **compose-url** is not set and contains a repo to be used for rpm module type testing (obsolote)
+* **container** contains a link to a container (docker.io or local tar.gz file) (obsolete)
+* **repo** is used when **compose-url** is not set and contains a repo to be used for rpm module type testing (obsolete)
 * **parent** if you would like to have more configs for same module type, it is possible to do it via inheritance. There will be used parent module + overwritten values with this one, you can rewrite whatever you want. You have to set parent (base) module type allowed are just **rpm/docker**
+* **template** contains a link to a OpenShift template. The template is added into OpenShift resources, like template and new application is created based on the template. If `template` is not specified, then new application is created based on container link, by command `oc new-app ...`
+* **docker_pull** specifies if image is pulled by command `docker pull` or not before adding to an OpenShift registry. Turning off does not overwrite your local image.
```
Comment (TomasTomecek) [inline docs/user_guide/how_to_write_conf_file.rst]: could be also worth specifying what values are users meant to type here

File: moduleframework/common.py [hunk 17]
```diff
@@ -457,10 +461,31 @@ def get_url(self):
...
+        """
+        return self.info.get("template")
+
+    def get_docker_pull(self):
+        """
+        Gets boolean value if image should be pulled or not.
+        :return: bool
+        """
+        docker_pull = self.info.get("docker_pull")
+        print_debug(docker_pull)
```
Comment (TomasTomecek) [inline moduleframework/common.py]: this will just print "true", or "0" right? Would be beneficial to prepend a message what the value represents.

File: moduleframework/helpers/openshift_helper.py [hunk 18]
```diff
@@ -54,8 +58,83 @@ def __init__(self):
...
+                                                            random_str=random_str)
         common.print_debug(self.icontainer, self.app_name)
 
+    def _get_openshift_ip_registry(self):
+        """
+        Function returns an IP of OpenShift registry.
+        Command which is used for getting this info is `oc get svc -n default docker-registry`
+        :return: str or None
+        """
+        openshift_ip_register = None
```
Comment (TomasTomecek) [inline moduleframework/helpers/openshift_helper.py]: nit: it's registry, not register

File: moduleframework/helpers/openshift_helper.py [hunk 19]
```diff
@@ -54,8 +58,83 @@ def __init__(self):
...
+        :param account: Either user specified account or 'system:admin'
+        :param password: Either user specified account
+        """
+        if password is None:
+            s = self.runHost("oc login -u %s" % account, verbose=common.is_not_silent())
+        else:
+            s = self.runHost("oc login -u %s -p %s" % (account, password), verbose=common.is_not_silent())
+
+
+    def _register_docker_to_openshift_register(self):
```
Comment (TomasTomecek) [inline moduleframework/helpers/openshift_helper.py]: registry

File: moduleframework/helpers/openshift_helper.py [hunk 20]
```diff
@@ -54,8 +58,83 @@ def __init__(self):
...
+        """
+        Function registers an OpenShift docker-registry into docker
```
Comment (TomasTomecek) [inline moduleframework/helpers/openshift_helper.py]: it actually just pushes the container image inside OpenShift's registry; you don't *register registry into docker*

File: moduleframework/helpers/openshift_helper.py [hunk 21]
```diff
@@ -54,8 +58,83 @@ def __init__(self):
...
+        * gets an IP of OpenShift registry
+        * switch to user defined OpenShift user and password.
+        * runs docker login with token and IP:5000
+        * tag container name
+        * push container name into docker
+        """
+        token = None
+        for i in range(0,5):
+            whoami = self.runHost("oc whoami -t", ignore_status=True, verbose=common.is_not_silent())
+            if whoami.exit_status == 0:
```
Comment (TomasTomecek) [inline moduleframework/helpers/openshift_helper.py]: do we workaround something here?

File: moduleframework/helpers/openshift_helper.py [hunk 22]
```diff
@@ -54,8 +58,83 @@ def __init__(self):
...
+        openshift_ip_register = self._get_openshift_ip_registry()
+        self._change_openshift_account(account=common.get_openshift_user(),
+                                password=common.get_openshift_passwd())
+
+        for i in range(0,5):
+            docker_login = self.runHost('docker login -u {user} -p {token} {ip}:5000'.format(
+                user=common.get_openshift_user(),
+                token=token,
+                ip=openshift_ip_register), ignore_status=True, verbose=common.is_not_silent())
+            if docker_login.exit_status == 0:
```
Comment (TomasTomecek) [inline moduleframework/helpers/openshift_helper.py]: once more, what do we workaround here?

File: moduleframework/common.py (hunk 17 above)
Comment (phracek) [reply moduleframework/common.py]: This is typo. It will be deleted.

File: moduleframework/helpers/openshift_helper.py (hunk 21 above)
Comment (phracek) [reply moduleframework/helpers/openshift_helper.py]: For some reason it sometimes failed. Some time it expires or I don't get a token. I will check it more deeply. The same as below.

File: docs/user_guide/how_to_write_conf_file.rst [hunk 23]
```diff
@@ -62,16 +66,18 @@ An example of module types specification:
...
 * **labels** contains docker labels to check if there is any
 * **url** contains link to a container or repo (same meaning as container or repo)
-* **container** contains a link to a container (docker.io or local tar.gz file) (obsolote)
-* **repo** is used when **compose-url** is not set and contains a repo to be used for rpm module type testing (obsolote)
+* **container** contains a link to a container (docker.io or local tar.gz file) (obsolete)
+* **repo** is used when **compose-url** is not set and contains a repo to be used for rpm module type testing (obsolete)
 * **parent** if you would like to have more configs for same module type, it is possible to do it via inheritance. There will be used parent module + overwritten values with this one, you can rewrite whatever you want. You have to set parent (base) module type allowed are just **rpm/docker**
+* **template** contains a link to a OpenShift template. The template is added into OpenShift resources, like template and new application is created based on the template. If `template` is not specified, then new application is created based on container link, by command `oc new-app ...`
```
[the remark this answers is not in the document]
Comment (phracek) [reply docs/user_guide/how_to_write_conf_file.rst]: well, It can be either URL link or a path. I will rephrase it to `contains an URL link or a path to an OpenShift template.`
